### ISSUE TITLE
Mismatched signature in trait implementation with lifetimes not caught

### ISSUE THREAD

**Issue by Ms2ger** [rust-lang/rust#27083]
I *think* this shouldn't compile, but it does (in rustc 1.3.0-dev (fddfd089b 2015-07-10)):

```rust
struct Foo;

trait Bar {
    fn get_style<'a>(&'a self, r: &'a Foo) -> &'a Foo;
}

impl Bar for Foo {
    fn get_style<'a>(&self, r: &'a Foo) -> &'a Foo {
        r
    }
}

fn main() {}
```

**Comment by arielb1** [comment]
@Ms2ger

Could you extend it into a real unsoundness? I don't think this is a real problem: the trait type is `for<'a> fn(&'a Foo, &'a Foo) -> &'a Foo`, which is a subtype of the impl type `for<'a,'s> fn(&'s Foo, &'a Foo) -> &'a Foo` - it requires both of the argument lifetimes to be the same as the output lifetime, not just the second argument's lifetime, and we allow impl-types to be subtypes of the trait-type (e.g. you can even replace the return type with `&'static Foo` - of course only if you also adjust `get_style`'s body - and it would still work).

**Comment by Ms2ger** [comment]
I'm not sure anything's unsound here; I found this pattern in Servo and was surprised that it built at all, given the different signatures.

**Comment by arielb1** [comment]
I'll close this then.
